messaging/src/main/java/org/axonframework/messaging/deadletter/InMemorySequencedDeadLetterQueue.java

```diff
@@ -0,0 +1,378 @@
+                }
+            }
+        } else if (logger.isDebugEnabled()) {
+            logger.debug("Cannot evict [{}] as it could not be found in this queue.", letter);
+        }
+    }
+
+    @Override
+    public void requeue(
+            @Nonnull DeadLetter<? extends M> letter,
+            @Nonnull UnaryOperator<DeadLetter<? extends M>> letterUpdater
+    ) throws NoSuchDeadLetterException {
+        Optional<Map.Entry<String, Deque<DeadLetter<? extends M>>>> optionalSequence =
+                deadLetters.entrySet()
+                           .stream()
+                           .filter(sequence -> sequence.getValue().remove(letter))
+                           .findFirst();
+
+        if (optionalSequence.isPresent()) {
+            synchronized (deadLetters) {
+                String sequenceId = optionalSequence.get().getKey();
+                deadLetters.get(sequenceId)
+                           .addFirst(letterUpdater.apply(letter.markTouched()));
+                if (logger.isTraceEnabled()) {
+                    logger.trace("Requeued letter [{}] for sequence [{}].", letter, sequenceId);
+                }
+            }
+        } else {
+            throw new NoSuchDeadLetterException(
+                    "Cannot requeue [" + letter + "] since there is not matching entry in this queue."
+            );
+        }
+    }
+
+    @Override
+    public boolean contains(@Nonnull Object sequenceIdentifier) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("Validating existence of sequence identifier [{}].", sequenceIdentifier);
+        }
+        return contains(toIdentifier(sequenceIdentifier));
+    }
+
+    @Override
+    public Iterable<DeadLetter<? extends M>> deadLetterSequence(@Nonnull Object sequenceIdentifier) {
+        String identifier = toIdentifier(sequenceIdentifier);
+        return contains(identifier) ? new ArrayList<>(deadLetters.get(identifier)) : Collections.emptyList();
+    }
+
+    private boolean contains(String identifier) {
+        synchronized (deadLetters) {
+            return deadLetters.containsKey(identifier);
+        }
+    }
+
+    @Override
+    public Iterable<Iterable<DeadLetter<? extends M>>> deadLetters() {
+        return new ArrayList<>(deadLetters.values());
+    }
+
+    @Override
+    public boolean isFull(@Nonnull Object sequenceIdentifier) {
+        String identifier = toIdentifier(sequenceIdentifier);
+        return maximumNumberOfSequencesReached(identifier) || maximumSequenceSizeReached(identifier);
+    }
+
+    private static String toIdentifier(Object sequenceIdentifier) {
+        return sequenceIdentifier instanceof String
+                ? (String) sequenceIdentifier
+                : Integer.toString(sequenceIdentifier.hashCode());
+    }
+
+    private boolean maximumNumberOfSequencesReached(String identifier) {
+        return !deadLetters.containsKey(identifier) && deadLetters.keySet().size() >= maxSequences;
```

Comment:
Did some investigation and found that the order is correct but using a ConcurrentHashMap should improve performance since we are not using the extra features the ConcurrentSkipListMap brings

Reply:
I think we can benefit of the `ConcurrentSkipListMap`, as there's a chance that process letters from a sequence may happen while the regular event handling process inserts entries into the sequence.
At least, by looking at the [JavaDoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentSkipListMap.html), I anticipate the line "Insertion, removal, update, and access operations safely execute concurrently by multiple threads." ensures the above process to be safe.

Reply:
Of course...key is the `ConcurrentLinkedDeque` that becomes the sequence within the `deadLetters` map. 
I get both your points now, thanks. I'll adjust the `deadLetters` collection to a `ConcurrentHashMap`.